src/creds/creds.c

```diff
@@ -60,63 +60,72 @@ static const char* transcode_mode_table[_TRANSCODE_MAX] = {
 
 DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING(transcode_mode, TranscodeMode);
 
-static int open_credential_directory(DIR **ret) {
-        _cleanup_free_ char *j = NULL;
+static int open_credential_directory(
+                DIR **ret_dir,
+                const char **ret_prefix,
+                bool encrypted) {
+
         const char *p;
         DIR *d;
         int r;
 
-        if (arg_system) {
-                _cleanup_free_ char *cd = NULL;
-
-                r = getenv_for_pid(1, "CREDENTIALS_DIRECTORY", &cd);
+        assert(ret_dir);
+
+        if (arg_system)
+                /* PID 1 creates a fixed path symlinks for system credentials */
+                p = encrypted ?
+                        ENCRYPTED_SYSTEM_CREDENTIALS_DIRECTORY :
+                        SYSTEM_CREDENTIALS_DIRECTORY;
+        else {
+                /* Otherwise take the dirs from the env vars we got passed */
+                r = (encrypted ? get_encrypted_credentials_dir : get_credentials_dir)(&p);
+                if (r == -ENXIO) /* No environment variable? */
+                        goto not_found;
                 if (r < 0)
-                        return r;
-                if (!cd)
-                        return -ENXIO;
-
-                if (!path_is_absolute(cd) || !path_is_normalized(cd))
-                        return -EINVAL;
+                        return log_error_errno(r, "Failed to get credentials directory: %m");
+        }
 
-                j = path_join("/proc/1/root", cd);
-                if (!j)
-                        return -ENOMEM;
+        d = opendir(p);
+        if (!d) {
+                if (arg_system && errno == ENOENT)  /* No such dir? Then no creds where passed. */
```

Comment:
Why conditionalize this on arg_system? Is the directory guaranteed to be there for service specific directories?

Reply:
the per-service stuff relies on an env var, the per-system stuff uses hardcode path. if the env var points to non-existing path something is weird, and this should be considered an error. if the fixed path in the system case does not exist that's however totally ok, it just means nothing was passed.

will extend comment.